### DISCUSSION TITLE
`error(gtk_surface): surface failed to realize: Failed to create EGL display` on Intel iGPU Linux Mint

### DISCUSSION THREAD

**Discussion by alapins** [ghostty-org/ghostty#3763] [Q&A] [answered]
I'm on Linux Mint 22 (which is based on Ubuntu 24.04), with GTK 4.16.3 and libadwaita 1.6.2, running Cinnamon desktop env.  I do NOT have an nVidia graphics card; I'm using the iGPU on an Intel 14700.  

I tried compiling both locally using the nix build instructions from the docs, and also using `nix profile install github:ghostty-org/ghostty`. Compile was successful both ways.

Launching ghostty results in a new window with the message `Failed to create EGL display` and the logs below. The other discussions I've seen reference specific nVidia driver issues or logs indicating problems with OpenGL.  I've tried `GDK_DEBUG=gl-disable-gles ghostty` and  `LIBGL_ALWAYS_SOFTWARE=true ghostty` with no change. 

This seems to be a common issue for those running Ubuntu or one of it's derivatives, but I haven't seen anything that solves the issue (I'm running a supported version of GTK, not using nVidia -- not sure how to further debug the root cause)

```
info: ghostty version=1.0.1-bee2188-nix
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale after unset lang result=C
warning(os): setlocale failed with LANG and system default. Falling back to en_US.UTF-8
error(os): setlocale failed even with the fallback, uncertain results
info(gtk): GTK version build=4.16.3 runtime=4.16.3
info: optional config file not found, not loading path=/home/alex/.config/ghostty/config
info(config): default shell source=env value=/usr/bin/zsh
warning(sentry): crash report written to disk path=/home/alex/.local/state/ghostty/crash/9b656c7f-a85e-4139-468a-7340d89e4ce3.ghosttycrash
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2

(process:168963): Adwaita-WARNING **: 11:53:37.263: Using GtkSettings:gtk-application-prefer-dark-theme with libadwaita is unsupported. Please use AdwStyleManager:color-scheme instead.

(ghostty:168963): Gtk-WARNING **: 11:53:37.327: No IM module
[9b656c7f-a85e-4139-468a-7340d89e4ce3.ghosttycrash.zip](https://github.com/user-attachments/files/18266900/9b656c7f-a85e-4139-468a-7340d89e4ce3.ghosttycrash.zip)
 matching GTK_IM_MODULE=ibus found
error(gtk_surface): surface failed to realize: Failed to create EGL display
warning(gtk_surface): this error is usually due to a driver or gtk bug
warning(gtk_surface): this is a common cause of this issue: https://gitlab.gnome.org/GNOME/gtk/-/issues/4950
zsh: segmentation fault (core dumped)  ghostty
```
EDIT: Summary
----------------------
As identified by @ipsavitsky [here](https://github.com/ghostty-org/ghostty/discussions/3763#discussioncomment-11699970) this) is due to a problem with how OpenGL works with nix on anything other than NixOS.  There are a couple of solutions, depending on what you're doing with ghostty and how you use nix:

# You're evaluating Ghostty
As noted by @Nealium and @timmytandian  [here](https://github.com/ghostty-org/ghostty/discussions/3763#discussioncomment-11707119), 
`nix run github:nix-community/nixGL#nixGLIntel -- ghostty`
(You may need to replace nixGLIntel with the proper name for your graphics card, see [nixgl](https://github.com/nix-community/nixGL) for what wrappers are avaiialble.) 

This is awesome, because you're just immediately running ghostty, and don't have to worry about uninstalling it if you don't like it.

Run `nix-collect-garbage` to remove nixGL and ghostty entirely.

# Permanent install

## Home Manager
This is the easiest method, and requires knowing the least about nix vagaries.  Add the following to your home.nix:
```
{ config, lib, pkgs, ... }:

{
  nixGL.packages = import <nixgl> { inherit pkgs; };
  nixGL.defaultWrapper = "mesa";
  nixGL.installScripts = [ "mesa" ];

  home.packages = [
    (config.lib.nixGL.wrap pkgs.ghostty)
  ];
}
```
This will automatically wrap any runs of ghostty in nixgl, and the desktop shortcuts and running from the command line work as you'd expect, without further modification. See the [home manager GPU docs](https://nix-community.github.io/home-manager/index.xhtml#sec-usage-gpu-non-nixos) for more.

@Nealium with help from @frobware used home manager to override the desktop laucher [here](https://github.com/ghostty-org/ghostty/discussions/3763#discussioncomment-11701596), but this wouldn't address launching from the command line (and if the desktop definition changes, ex. new icon, it would have to be manually sync'ed)

## Flake/`nix profile`
```nix profile install github:nix-community/nixGL#nixGLIntel
nix profile install github:ghostty-org/ghostty
``` 
To run from terminal: `nixGLIntel ghostty`  
To run from launcher, you'll either have to:
* modify the desktop shortcut that's installed at `~/.nix-profile/share/applications/com.mitchellh.ghostty.desktop` and change `Exec=ghostty` to `Exec=nixGLIntel ghostty`.  You'll need to do this every time you update the install (using `nix profile upgrade`) :frowning_face: 
* create a new desktop file with a different name that contains the mods above.  This won't need to be changed.

## config.nix
It seems like this should be possible to do similarly to how home manager does it (the wrapper syntax seems to be based on the wrappers supported in straight nix) but I haven't tried doing it.

**Comment by jc00ke** [comment]
I'm seeing this as well on Debian Testing with stock GNOME 47.2 on an Intel UHD 620 integrated graphics card (this is an old Dell XPS 13).

**Comment by timmytandian** [comment]
I haven't found solution yet, just want to report that when I launch ghostty, a new window pop up with message "failed to create EGL display".  Similar outcome as alapins, although my system is slightly different.

**My system**:
- PoP!_OS based on Ubuntu 22.04
- No graphics card, only integrated GPU
- AMD CPU
- Used the nix build instructions from the docs to build Ghostty, had no problem with the compile.


**Ghostty version information**:
```
$ ghostty --version
Ghostty 1.0.1-87bd0bb-nix

Version
  - version: 1.0.1-87bd0bb-nix
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.16.3
    runtime    : 4.16.3
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
  - libX11     : enabled
```


Running ghostty returned error related to GTK. **Output**:
```
$ ghostty
info: ghostty version=1.0.1-87bd0bb-nix
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=LC_CTYPE=en_US.UTF-8;LC_NUMERIC=en_GB.UTF-8;LC_TIME=en_GB.UTF-8;LC_COLLATE=en_US.UTF-8;LC_MONETARY=en_GB.UTF-8;LC_MESSAGES=en_US.UTF-8;LC_PAPER=en_GB.UTF-8;LC_NAME=en_GB.UTF-8;LC_ADDRESS=en_GB.UTF-8;LC_TELEPHONE=en_GB.UTF-8;LC_MEASUREMENT=en_GB.UTF-8;LC_IDENTIFICATION=en_GB.UTF-8
info(gtk): GTK version build=4.16.3 runtime=4.16.3
info: optional config file not found, not loading path=/home/poptimmy/.config/ghostty/config
info(config): default shell source=env value=/usr/bin/zsh
warning(sentry): crash report written to disk path=/home/poptimmy/.local/state/ghostty/crash/7f2e0a5f-af84-4242-c67b-eeedabde9345.ghosttycrash
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2

(process:73404): Gtk-WARNING **: 19:58:57.257: Theme parser error: gtk.css:7917:18-22: Expected a number

(process:73404): Gtk-WARNING **: 19:58:57.258: Theme parser error: gtk.css:7929:18-22: Expected a number
error(gtk): unable to get current color scheme: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such method “ReadOne”

(ghostty:73404): Gtk-WARNING **: 19:58:57.417: No IM module matching GTK_IM_MODULE=ibus found
error(gtk_surface): surface failed to realize: Failed to create EGL display
warning(gtk_surface): this error is usually due to a driver or gtk bug
warning(gtk_surface): this is a common cause of this issue: https://gitlab.gnome.org/GNOME/gtk/-/issues/4950
zsh: segmentation fault (core dumped)  ghostty
```

**Comment by ipsavitsky** [comment, marked as answer]
I have encountered the same error, It's a quirk of how nix installations work on systems other than nixos. There exists a wrapper called [nixgl](https://github.com/nix-community/nixGL) that fixes this. Run with `nixGL ghostty`

**Reply by Nealium** [reply]
Perfect! That worked for me! Now to actually set that up

#### Test Command
`nix run github:nix-community/nixGL#nixGLIntel -- ghostty`

#### Sys
* Pop! OS (not cosmic)
* Ryzen 7
* AMD Radeon graphics

---
## Edit

I ended up downloading nixGL using the channels way, which got the `nixGL ghostty` command working. Then per @frobware 's direction, [Home Manager desktopEntries (options)](https://nix-community.github.io/home-manager/options.xhtml#opt-xdg.desktopEntries) and [xdg-desktop-entries.nix (Repo)](https://github.com/nix-community/home-manager/blob/master/modules/misc/xdg-desktop-entries.nix) added a desktopEntries item. I used the example below and, for me, it overwrites the default one
```nix
{ config, pkgs, ghostty, ... }:
let
  username = "neal";
in
{
  # ..home stuff

  home.packages = [
    ghostty.packages.x86_64-linux.default
  ]

  xdg.desktopEntries."com.mitchellh.ghostty" = {
    name = "Ghostty";
    type = "Application";
    comment = "A terminal emulator";
    exec = "nixGL ghostty";
    icon = "com.mitchellh.ghostty";
    terminal = false;
    startupNotify = true;
    categories = [ "System" "TerminalEmulator" ];
    settings = {
      Keywords = "terminal;tty;pty;";
      X-GNOME-UsesNotifications = "true";
      X-TerminalArgExec = "-e";
      X-TerminalArgTitle = "--title=";
      X-TerminalArgAppId = "--class=";
      X-TerminalArgDir = "--working-directory=";
      X-TerminalArgHold = "--wait-after-command";
    };
    actions = {
      new-window = {
        name = "New Window";
        exec = "nixGL ghostty";
      };
    };
  };
```
> Time to play! :partying_face:

**Reply by timmytandian** [reply]
Yes, I confirm @Nealium 's way works in my machine as well. Thank you @Nealium, @ipsavitsky , @frobware !
1. Install *nixGLIntel* using the channel method (`nix-env -iA nixgl.nixGLIntel`).
2. Make sure that `nix run github:nix-community/nixGL#nixGLIntel -- ghostty` works
3. Update my home.nix to configure ghostty and the desktop entry similar to Neal's code above. The difference is instead of `nixGL ghostty`, I run with `nixGLIntel ghostty`

**Reply by Nealium** [reply]
Thanks! @timmytandian Good change! I've updated by own to use `nixGLIntel` and not the default auto-detect.

I kept I wondering why `nix-env -iA nixgl.auto.nixGLIntel` wouldn't install. *I replaced "nixGLDefault" with my desired wrapper! :man_facepalming:*
> default can be uninstalled with: `nix-env -uA nixgl.auto.nixGLDefault`

#### Nix tidbit

As this might come in handy in the future, your 1 & 2 should be swapped:
because `nix run github:{x} -- {command}` is pulling directly from github, you don't have to fully commit to installing it to the system through channels or flakes. Nix will install it in a temporary environment and run it. Then It'll be auto-deleted the next time you run `nix-collect-garbage`. Quite nice for quickly trying out a *potential* one-off program without bogging down your system!
cheers! ☕

**Comment by frobware** [comment]
I do something similar for Alacritty; a snippet from my config:

```nix
{ config, lib, pkgs, ... }:

let
  customAlacrittyDesktopEntry = {
    name = "Alacritty";
    genericName = "Terminal";
    exec = "nixGLIntel alacritty";
    icon = "Alacritty";
    terminal = false;
    categories = [ "System" "TerminalEmulator" ];
    comment = "A fast, cross-platform, OpenGL terminal emulator";
    actions = {
      New = {
        name = "New Terminal";
        exec = "alacritty-new-window";
      };
    };
  };
...
```

**Reply by alapins** [reply]
Thanks very much @ipsavitsky, @Nealium, @frobware & @timmytandian!  Confirmed that this works on my system, also using the nixGLIntel.

**Comment by jc00ke** [comment]
Thanks everyone, this set me down the right path! I ended up wrapping `ghostty` with `nixGL` like the [home-manager docs](https://nix-community.github.io/home-manager/index.xhtml#sec-usage-gpu-non-nixos) suggest:

```nix
   nixGL.packages = import <nixgl> { inherit pkgs; };
   nixGL.defaultWrapper = "mesa";
   nixGL.offloadWrapper = "intel";

  programs.ghostty = {
    enable = true;
    package = config.lib.nixGL.wrap pkgs.ghostty;
  };
```

Using that method, my `.desktop` entry was just `ghostty` instead of `nixGLIntel ghostty`. I had to use [this workaround](https://discourse.nixos.org/t/apps-installed-via-home-manager-are-not-visible-within-gnome/48252/2) to get the desktop entry to show up in GNOME, and when it did the icon was generic... I'm still working on that.

<details><summary>Narrator: he worked it out</summary>
I needed to modify the desktop entries workaround to also symlink the icons directory:

```bash
    if [ "$XDG_CURRENT_DESKTOP" = "GNOME" ]; then

      mkdir -p "${config.home.homeDirectory}/.local/share/applications"

      if [ -d "${config.home.homeDirectory}/.local/share/applications/nix" ]; then
        rm -rf "${config.home.homeDirectory}/.local/share/applications/nix"
      fi

      ln -sf "${config.home.homeDirectory}/.nix-profile/share/applications" \
        ${config.home.homeDirectory}/.local/share/applications/nix

      mkdir -p "${config.home.homeDirectory}/.local/share/icons"

      if [ -d "${config.home.homeDirectory}/.local/share/icons/nix" ]; then
        rm -rf "${config.home.homeDirectory}/.local/share/icons/nix"
      fi

      ln -sf "${config.home.homeDirectory}/.nix-profile/share/icons" \
        ${config.home.homeDirectory}/.local/share/icons/nix
    fi
```
</details>

[Here is my most recent working home-manager config](https://github.com/jc00ke/dotfiles/commit/3943ec4061ab85527a0eeeb6ecd193debebcdeae), if that helps anyone out.

**Reply by jc00ke** [reply]
[Here's an update](https://github.com/jc00ke/dotfiles/commit/6e255a0137219f60ca1493af0128bf3e69ac4846) I made to get it working on my work desktop with a Radeon GPU too.

**Comment by ndom91** [comment]
Anyone figured out a way to do this without `NixGL`? Adding nixgl balloons the diff of my `switch` output to include an additional copy of (all?) opengl apps (?) I'm unsure what the common denominator of all these extra applications is, but they do all seem graphics-heavy.


<img width="1014" height="966" alt="image" src="https://github.com/user-attachments/assets/9cad83c3-ccc9-49ae-ba2e-1d455cf9ec09" />

<img width="1014" height="966" alt="image" src="https://github.com/user-attachments/assets/7041c118-8fdf-469b-b037-27ef2da4874a" />
